rfc/2022/2022-11-02_TechTeam4_Preview_Environment_Architecture.md

```diff
@@ -0,0 +1,97 @@
+
+For the MVP of the preview environment architecture, the user will interact with GitHub normally and the preview environment will be created with default settings. As a developer pushes their branch to `vets-website` or `vets-api`, the preview environment architecture will detect the branch and begin the process of deploying the code in the branch to a preview environment. Once the continuous integration workflow completes successfully, a request to deploy the preview environment triggers a GitHub Actions workflow. The GHA workflow writes the dynamically generated preview environment configuration to the `app-manifest` repository. EKS detects the change in the `devops` repository and deploys the newly defined environment.
+
+A diagram is submitted [below](#Diagrams) as additional explanation.
+
+**MVP Feature Set:**
+1. Preview Environment MVP will create ephemeral environments that can be accessed without SOCKS
+2. Preview Environment MVP will operate on the EKS infrastructure and paradigm
+3. Preview Environment MVP will automatically create a Preview Environment when a PR is created
+4. Preview Environment MVP will utilize mock test data
+
+### Post-MVP Design: Additional Features
+For reference, there are numerous additional features planned for inclusion in Post-MVP iterations. Specifically, we will begin adding the abilities to specify custom settings like: custom repository configurations, selection of different types of test data, connectivity to lower environment endpoints for external partners. These features will be vetted and prioritized as development progresses.
+
+## Risks
+* Removing the ability for a preview environment to integrate with VA enterprise lower environments may initially reduce the quality of the experience of preview environments for some users. These risks will lessen over time as additional functionality is added into preview environments and this capability is restored.
+* If there is an overlap between the creation of review instances and the creation of preview environments there could be increased resource usage and cost while both products are supported.
+
+<!--
+List the risks of this approach
+
+* There are always risks. What are the risks of this solution?
+* These are the things people will bring up in opposition to your idea or plans. Acknowledge them.
+-->
+## Alternatives
+There were a number of approaches explored during Discovery, listed first below is the recommended approach, followed by the two other primary solutions considered. Lastly, we included a feature comparison to inform everyone on what led into our decision making process.
+
+### Custom preview environment build process: (Selected Approach)
```

Comment:
How does this approach remove SOCKS requirement?

Reply:
Our understanding from discovery is that two characteristics in combination are the current cause for the SOCKS requirement with Review Instances --
1. Review instances allow users who can create branches in `vets-website` or `vets-api` to create an instance of VA.gov that runs arbitrary code as long as it passes CI
2. Review instances are configured to integrate with lower environments within the VA enterprise through a proxy configuration that sends traffic through the staging `vets-api` instance.

The new approach for preview environments removes the integration with lower environments from in the MVP, and instead relies on mocked data available in `vets-api-mockdata`. In a subsequent release, we plan to enable the user to configure their preview environment to integrate with lower environments instead. As part of that configuration, the preview environment could also be set to authN / authZ the users hitting the preview environment.

Reply:
What VPC would preview environments exist in, how would the DNS be handled?

Reply:
We expect that for MVP we will create PEs in the `vagov-dev` VPC. There are two options for DNS that we understand --
1. Create a sub-domain, something like `preview.va.gov`, and use AWS Route53 to configure traffic to reach resources in the dev VPC
2. Append paths to an existing domain for each PE and update the revproxy to route that path to the PE each time

We would prefer option 1, but we understand the creation of a sub-domain can be a lengthy and difficult process.

Reply:
Ok for 1) would that mean provisioning DNS entries for subdomains like feature-branch-foo.preview.vfs.va.gov?
If so it should be verified if this can be done through AWS or must be performed at the TIC.

Can we get a network diagram for one or both of these options, and any potential iteration that would connect to non-mocked external systems?

Reply:
To-do: 
- [ ] What expectations do we have around ESECC requests?
  - [ ] Do we expect the WASA scan to apply to our environments as well?
    - [ ] Expectation / risk in assuming that WASA scan previously executed will apply to our preview environments
      - ESECC controls DNS records for va doman / sub-domain
  - [ ] How are we approaching access controls?
  - [ ] Include network diagram for MVP solution
    - [ ] Include path delineation details in network diagram
  - [ ] Include network diagram for post-MVP solution
  - [ ] Include discussion of trade-offs from MVP and future post-MVP feature sets

Reply:
Reviewing the tradeoffs from MVP and future post-MVP feature sets, it’s helpful to first understand the planned feature sets of both, and then the delta between the two.
**MVP features will include:**
- Provide users the ability to create an ephemeral environment that enables the team to test and demo their product changes
- Provide the ability to share these ephemeral environments publicly with stakeholders and Veteran users for testing
- Provide users the ability to create preview environments that contain their product changes combined with the latest version of the VA.gov Platform
- Provide users the ability to to use mock data for testing and demoing their products
- Create preview environments using the EKS deployment paradigm
- Provide users with preview environments that are automatically created when a PR is committed

**Post-MVP features will include: (assuming time/schedule/priorities allow for all features to be developed)**
- Provide users the ability to create an ephemeral environment that enables the team to test and demo their product changes
- Provide the ability to share these ephemeral environments publicly with stakeholders and Veteran users for testing
- Provide users the ability to create preview environments that contain the exact changes they need from the front-end and back-end of their product
- they need from the front-end and back-end of their product
Provide users the ability to select from three categories of test data based on their use case – mocks, test harnesses, integration with lower environments
- Create preview environments using the EKS deployment paradigm
- Enable users to specify restricted access to a preview environment